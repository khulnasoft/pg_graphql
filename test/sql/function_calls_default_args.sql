begin;

    create function both_args_optional(
        a smallint default null,
        b integer default null
    )
        returns text language plpgsql immutable
    as $$
    begin

        if a is null and b is null then
            return 'both null';
        end if;

        if a is null then
            return 'b = ' || b::text;
        end if;

        if b is null then
            return 'a = ' || a::text;
        end if;

        return 'a = ' || a::text || ', b = ' || b::text;
    end;
    $$;

    select jsonb_pretty(graphql.resolve($$
        query {
            bothArgsOptional
        }
    $$));

    select jsonb_pretty(graphql.resolve($$
        query {
            bothArgsOptional(a: 1)
        }
    $$));

    select jsonb_pretty(graphql.resolve($$
        query {
            bothArgsOptional(b: 2)
        }
    $$));

    select jsonb_pretty(graphql.resolve($$
        query {
            bothArgsOptional(a: 1, b: 2)
        }
    $$));

    create function func_with_null_defaults(
        a smallint default null,
        b int default null,
        c bigint default null,
        d real default null,
        e double precision default null,
        f numeric default null,
        g bool default null,
        h uuid default null,
        i text default null,
        j date default null,
        k time default null,
        l time with time zone default null,
        m timestamp default null,
        n timestamptz default null,
        o json default null,
        p jsonb default null,
        q char(2) default null,
        r varchar(2) default null
    )
        returns text language plpgsql immutable
    as $$
    begin

        if
            a is null and
            b is null and
            c is null and
            d is null and
            e is null and
            f is null and
            g is null and
            h is null and
            i is null and
            j is null and
            k is null and
            l is null and
            m is null and
            n is null and
            o is null and
            p is null and
            q is null and
            r is null
        then
            return 'all args null';
        end if;

        return
            a::text || ', ' ||
            b::text || ', ' ||
            c::text || ', ' ||
            d::text || ', ' ||
            e::text || ', ' ||
            f::text || ', ' ||
            g::text || ', ' ||
            h::text || ', ' ||
            i::text || ', ' ||
            j::text || ', ' ||
            k::text || ', ' ||
            l::text || ', ' ||
            m::text || ', ' ||
            n::text || ', ' ||
            o::text || ', ' ||
            p::text || ', ' ||
            q::text || ', ' ||
            r::text;
    end;
    $$;

    select jsonb_pretty(graphql.resolve($$
        query {
            funcWithNullDefaults
        }
    $$));

    select jsonb_pretty(graphql.resolve($$
        query {
            funcWithNullDefaults(
                a: 1,
                b: 2,
                c: 3,
                d: 1.1,
                e: 2.3,
                f: "3.4",
                g: false,
                h: "8871277a-de9c-4156-b31f-5b4060001081",
                i: "text",
                j: "2023-07-28",
                k: "10:20",
                l: "10:20+05:30",
                m: "2023-07-28 12:39:05",
                n: "2023-07-28 12:39:05+05:30",
                o: "{\"a\": {\"b\": \"foo\"}}",
                p: "{\"a\": {\"b\": \"foo\"}}",
                q: "hi",
                r: "ho"
            )
        }
    $$));

    create function defaul_null_mixed_case(
        a smallint default null,
        b integer default NULL,
        c integer default NuLl
    )
        returns text language plpgsql immutable
    as $$
    begin
        return 'mixed case';
    end;
    $$;

    select jsonb_pretty(
        graphql.resolve($$
            query IntrospectionQuery {
              __schema {
                queryType {
                  name
                  fields {
                    name
                    args {
                      name
                      defaultValue
                      type {
                        kind
                        name
                        ofType {
                            kind
                            name
                        }
                      }
                    }
                  }
                }
              }
            }
        $$)
    );

rollback;
