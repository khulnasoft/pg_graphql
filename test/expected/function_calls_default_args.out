begin;
    create function both_args_optional(
        a smallint default null,
        b integer default null
    )
        returns text language plpgsql immutable
    as $$
    begin

        if a is null and b is null then
            return 'both null';
        end if;

        if a is null then
            return 'b = ' || b::text;
        end if;

        if b is null then
            return 'a = ' || a::text;
        end if;

        return 'a = ' || a::text || ', b = ' || b::text;
    end;
    $$;
    select jsonb_pretty(graphql.resolve($$
        query {
            bothArgsOptional
        }
    $$));
              jsonb_pretty               
-----------------------------------------
 {                                      +
     "data": {                          +
         "bothArgsOptional": "both null"+
     }                                  +
 }
(1 row)

    select jsonb_pretty(graphql.resolve($$
        query {
            bothArgsOptional(a: 1)
        }
    $$));
            jsonb_pretty             
-------------------------------------
 {                                  +
     "data": {                      +
         "bothArgsOptional": "a = 1"+
     }                              +
 }
(1 row)

    select jsonb_pretty(graphql.resolve($$
        query {
            bothArgsOptional(b: 2)
        }
    $$));
            jsonb_pretty             
-------------------------------------
 {                                  +
     "data": {                      +
         "bothArgsOptional": "b = 2"+
     }                              +
 }
(1 row)

    select jsonb_pretty(graphql.resolve($$
        query {
            bothArgsOptional(a: 1, b: 2)
        }
    $$));
                jsonb_pretty                
--------------------------------------------
 {                                         +
     "data": {                             +
         "bothArgsOptional": "a = 1, b = 2"+
     }                                     +
 }
(1 row)

    create function func_with_null_defaults(
        a smallint default null,
        b int default null,
        c bigint default null,
        d real default null,
        e double precision default null,
        f numeric default null,
        g bool default null,
        h uuid default null,
        i text default null,
        j date default null,
        k time default null,
        l time with time zone default null,
        m timestamp default null,
        n timestamptz default null,
        o json default null,
        p jsonb default null,
        q char(2) default null,
        r varchar(2) default null
    )
        returns text language plpgsql immutable
    as $$
    begin

        if
            a is null and
            b is null and
            c is null and
            d is null and
            e is null and
            f is null and
            g is null and
            h is null and
            i is null and
            j is null and
            k is null and
            l is null and
            m is null and
            n is null and
            o is null and
            p is null and
            q is null and
            r is null
        then
            return 'all args null';
        end if;

        return
            a::text || ', ' ||
            b::text || ', ' ||
            c::text || ', ' ||
            d::text || ', ' ||
            e::text || ', ' ||
            f::text || ', ' ||
            g::text || ', ' ||
            h::text || ', ' ||
            i::text || ', ' ||
            j::text || ', ' ||
            k::text || ', ' ||
            l::text || ', ' ||
            m::text || ', ' ||
            n::text || ', ' ||
            o::text || ', ' ||
            p::text || ', ' ||
            q::text || ', ' ||
            r::text;
    end;
    $$;
    select jsonb_pretty(graphql.resolve($$
        query {
            funcWithNullDefaults
        }
    $$));
                  jsonb_pretty                   
-------------------------------------------------
 {                                              +
     "data": {                                  +
         "funcWithNullDefaults": "all args null"+
     }                                          +
 }
(1 row)

    select jsonb_pretty(graphql.resolve($$
        query {
            funcWithNullDefaults(
                a: 1,
                b: 2,
                c: 3,
                d: 1.1,
                e: 2.3,
                f: "3.4",
                g: false,
                h: "8871277a-de9c-4156-b31f-5b4060001081",
                i: "text",
                j: "2023-07-28",
                k: "10:20",
                l: "10:20+05:30",
                m: "2023-07-28 12:39:05",
                n: "2023-07-28 12:39:05+05:30",
                o: "{\"a\": {\"b\": \"foo\"}}",
                p: "{\"a\": {\"b\": \"foo\"}}",
                q: "hi",
                r: "ho"
            )
        }
    $$));
                                                                                                                              jsonb_pretty                                                                                                                               
-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
 {                                                                                                                                                                                                                                                                      +
     "data": {                                                                                                                                                                                                                                                          +
         "funcWithNullDefaults": "1, 2, 3, 1.1, 2.3, 3.4, false, 8871277a-de9c-4156-b31f-5b4060001081, text, 07-28-2023, 10:20:00, 10:20:00+05:30, Fri Jul 28 12:39:05 2023, Fri Jul 28 00:09:05 2023 PDT, {\"a\": {\"b\": \"foo\"}}, {\"a\": {\"b\": \"foo\"}}, hi, ho"+
     }                                                                                                                                                                                                                                                                  +
 }
(1 row)

    create function defaul_null_mixed_case(
        a smallint default null,
        b integer default NULL,
        c integer default NuLl
    )
        returns text language plpgsql immutable
    as $$
    begin
        return 'mixed case';
    end;
    $$;
    select jsonb_pretty(
        graphql.resolve($$
            query IntrospectionQuery {
              __schema {
                queryType {
                  name
                  fields {
                    name
                    args {
                      name
                      defaultValue
                      type {
                        kind
                        name
                        ofType {
                            kind
                            name
                        }
                      }
                    }
                  }
                }
              }
            }
        $$)
    );
                       jsonb_pretty                        
-----------------------------------------------------------
 {                                                        +
     "data": {                                            +
         "__schema": {                                    +
             "queryType": {                               +
                 "name": "Query",                         +
                 "fields": [                              +
                     {                                    +
                         "args": [                        +
                             {                            +
                                 "name": "a",             +
                                 "type": {                +
                                     "kind": "SCALAR",    +
                                     "name": "Int",       +
                                     "ofType": null       +
                                 },                       +
                                 "defaultValue": null     +
                             },                           +
                             {                            +
                                 "name": "b",             +
                                 "type": {                +
                                     "kind": "SCALAR",    +
                                     "name": "Int",       +
                                     "ofType": null       +
                                 },                       +
                                 "defaultValue": null     +
                             }                            +
                         ],                               +
                         "name": "bothArgsOptional"       +
                     },                                   +
                     {                                    +
                         "args": [                        +
                             {                            +
                                 "name": "a",             +
                                 "type": {                +
                                     "kind": "SCALAR",    +
                                     "name": "Int",       +
                                     "ofType": null       +
                                 },                       +
                                 "defaultValue": null     +
                             },                           +
                             {                            +
                                 "name": "b",             +
                                 "type": {                +
                                     "kind": "SCALAR",    +
                                     "name": "Int",       +
                                     "ofType": null       +
                                 },                       +
                                 "defaultValue": null     +
                             },                           +
                             {                            +
                                 "name": "c",             +
                                 "type": {                +
                                     "kind": "SCALAR",    +
                                     "name": "Int",       +
                                     "ofType": null       +
                                 },                       +
                                 "defaultValue": null     +
                             }                            +
                         ],                               +
                         "name": "defaulNullMixedCase"    +
                     },                                   +
                     {                                    +
                         "args": [                        +
                             {                            +
                                 "name": "a",             +
                                 "type": {                +
                                     "kind": "SCALAR",    +
                                     "name": "Int",       +
                                     "ofType": null       +
                                 },                       +
                                 "defaultValue": null     +
                             },                           +
                             {                            +
                                 "name": "b",             +
                                 "type": {                +
                                     "kind": "SCALAR",    +
                                     "name": "Int",       +
                                     "ofType": null       +
                                 },                       +
                                 "defaultValue": null     +
                             },                           +
                             {                            +
                                 "name": "c",             +
                                 "type": {                +
                                     "kind": "SCALAR",    +
                                     "name": "BigInt",    +
                                     "ofType": null       +
                                 },                       +
                                 "defaultValue": null     +
                             },                           +
                             {                            +
                                 "name": "d",             +
                                 "type": {                +
                                     "kind": "SCALAR",    +
                                     "name": "Float",     +
                                     "ofType": null       +
                                 },                       +
                                 "defaultValue": null     +
                             },                           +
                             {                            +
                                 "name": "e",             +
                                 "type": {                +
                                     "kind": "SCALAR",    +
                                     "name": "Float",     +
                                     "ofType": null       +
                                 },                       +
                                 "defaultValue": null     +
                             },                           +
                             {                            +
                                 "name": "f",             +
                                 "type": {                +
                                     "kind": "SCALAR",    +
                                     "name": "BigFloat",  +
                                     "ofType": null       +
                                 },                       +
                                 "defaultValue": null     +
                             },                           +
                             {                            +
                                 "name": "g",             +
                                 "type": {                +
                                     "kind": "SCALAR",    +
                                     "name": "Boolean",   +
                                     "ofType": null       +
                                 },                       +
                                 "defaultValue": null     +
                             },                           +
                             {                            +
                                 "name": "h",             +
                                 "type": {                +
                                     "kind": "SCALAR",    +
                                     "name": "UUID",      +
                                     "ofType": null       +
                                 },                       +
                                 "defaultValue": null     +
                             },                           +
                             {                            +
                                 "name": "i",             +
                                 "type": {                +
                                     "kind": "SCALAR",    +
                                     "name": "String",    +
                                     "ofType": null       +
                                 },                       +
                                 "defaultValue": null     +
                             },                           +
                             {                            +
                                 "name": "j",             +
                                 "type": {                +
                                     "kind": "SCALAR",    +
                                     "name": "Date",      +
                                     "ofType": null       +
                                 },                       +
                                 "defaultValue": null     +
                             },                           +
                             {                            +
                                 "name": "k",             +
                                 "type": {                +
                                     "kind": "SCALAR",    +
                                     "name": "Time",      +
                                     "ofType": null       +
                                 },                       +
                                 "defaultValue": null     +
                             },                           +
                             {                            +
                                 "name": "l",             +
                                 "type": {                +
                                     "kind": "SCALAR",    +
                                     "name": "Opaque",    +
                                     "ofType": null       +
                                 },                       +
                                 "defaultValue": null     +
                             },                           +
                             {                            +
                                 "name": "m",             +
                                 "type": {                +
                                     "kind": "SCALAR",    +
                                     "name": "Datetime",  +
                                     "ofType": null       +
                                 },                       +
                                 "defaultValue": null     +
                             },                           +
                             {                            +
                                 "name": "n",             +
                                 "type": {                +
                                     "kind": "SCALAR",    +
                                     "name": "Datetime",  +
                                     "ofType": null       +
                                 },                       +
                                 "defaultValue": null     +
                             },                           +
                             {                            +
                                 "name": "o",             +
                                 "type": {                +
                                     "kind": "SCALAR",    +
                                     "name": "JSON",      +
                                     "ofType": null       +
                                 },                       +
                                 "defaultValue": null     +
                             },                           +
                             {                            +
                                 "name": "p",             +
                                 "type": {                +
                                     "kind": "SCALAR",    +
                                     "name": "JSON",      +
                                     "ofType": null       +
                                 },                       +
                                 "defaultValue": null     +
                             },                           +
                             {                            +
                                 "name": "q",             +
                                 "type": {                +
                                     "kind": "SCALAR",    +
                                     "name": "String",    +
                                     "ofType": null       +
                                 },                       +
                                 "defaultValue": null     +
                             },                           +
                             {                            +
                                 "name": "r",             +
                                 "type": {                +
                                     "kind": "SCALAR",    +
                                     "name": "String",    +
                                     "ofType": null       +
                                 },                       +
                                 "defaultValue": null     +
                             }                            +
                         ],                               +
                         "name": "funcWithNullDefaults"   +
                     },                                   +
                     {                                    +
                         "args": [                        +
                             {                            +
                                 "name": "nodeId",        +
                                 "type": {                +
                                     "kind": "NON_NULL",  +
                                     "name": null,        +
                                     "ofType": {          +
                                         "kind": "SCALAR",+
                                         "name": "ID"     +
                                     }                    +
                                 },                       +
                                 "defaultValue": null     +
                             }                            +
                         ],                               +
                         "name": "node"                   +
                     }                                    +
                 ]                                        +
             }                                            +
         }                                                +
     }                                                    +
 }
(1 row)

rollback;
